Start timer for posts

ampeigonet opened the pull request:
Story: https://trello.com/c/ZRue92zs/5-start-a-timer-to-post-topics

Commit:
Add post your topics button and timer

Timer should be shown after clicking on "Post your topics!" button once .js is added

Commit:
Add timer logic

Commit:
Add dynamic <script> content in each view

There were errors in the home page when the retro scripts would be loaded since it wouldn't find some document elements.

fedekau commented on app/assets/javascripts/retros.js:
```diff
@@ -0,0 +1,63 @@
+    let continueTimer = checkTimeLeft();
+
+    if (!continueTimer) {
+      clearInterval();
```
This will not stop the interval. The `setInterval` function returns an `intervalId` which should be used to stop the interval, this is because you can have many intervals running at the same time. You can see the specs on https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/clearInterval

or you can test it by doing 

```js
const intervalId = setInterval(() => console.log("I'am running!", 2000);

clearInterval() // Will not stop the interval
clearInterval(intervalId); // Will stop the interval
```

fedekau commented on app/assets/javascripts/retros.js:
```diff
@@ -0,0 +1,63 @@
+        document.getElementById("time").innerText = newValue;
+      }
+      if (timeLeft === 0) {
+        clearInterval(timeLeft);
```
Same as above, in this case `timeLeft` is not the `ID` of the interval and thus this call will no stop it.

fedekau reviewed (commented):
Looks good in general 🎉 

General notes:

- You need to be aware you are using the most recent version of JS here, this is not supported by all browsers, maybe you can investigate how to transpile JS in Rails.
- If you use a linting tool like ESLint you will probably get warn about some style issues, not a big deal, but it is always nice to have that.

If you have other doubts about the JS code of this PR let me know, I will gladly help 😄

fchavat commented on app/views/layouts/application.html.erb:
```diff
@@ -7,10 +7,10 @@
     <%= csp_meta_tag %>
 
     <%= stylesheet_link_tag    'application', media: 'all', 'data-turbolinks-track': 'reload' %>
-    <%= javascript_include_tag 'application', 'data-turbolinks-track': 'reload' %>
```
`application.js` is the file that works as a [manifest](https://guides.rubyonrails.org/asset_pipeline.html#manifest-files-and-directives) by default for the Asset Pipeline. So, if you don't include it, the libraries/files required inside this manifest (rails-ujs, activestorage, etc) will be no longer included in the webpage.
Also, as you can see in `application.js`, there is a line defined as `require_tree .` that will recursively include all files starting from `.`

ampeigonet replied:
Related to 👇

fchavat commented on app/views/retros/new.html.erb:
```diff
@@ -1,11 +1,16 @@
 </div>
+
+<% content_for :script do %>
+  <%= javascript_include_tag 'retros' %>
```
Related to the other comment. If you don't include the main manifest, `rails-ujs` and `turbolinks` won't be required.
Is there a reason to stop including `application.js` and start to include the files separately?

ampeigonet replied:
The reason why I removed the `javascript_include_tag` in `application.html.erb` is because the following error was being displayed in the console when loading the main page:

<img width="573" alt="screen shot 2018-10-17 at 15 50 59" src="https://user-images.githubusercontent.com/8642789/47109400-8366ad00-d224-11e8-8191-188b87dbe38f.png">

This is because as you said, all the .js files are included in every page, therefore `retros.js` looks for elements that belong to `retros/new` in `/` and doesn't find them, so they are undefined even though the document is indeed ready.

Maybe there's another way to fix this issue ?

fchavat replied:
I think that the problem there is that if we are not in `/retros/new` then `startNewRetroButton` in `retros.js` is undefined. So maybe this error can be fixed asking if this variable is defined before adding an event listener to it.

Anyway, I think that there has to be a better way to organize the javascript code without being defensive and hacky. Maybe Rodri or Larry knows more about that

ampeigonet replied:
This will be solved in https://github.com/anita040593/RetroApp/pull/6 because I'll add `<%= javascript_include_tag 'application', 'data-turbolinks-track': 'reload' %>` again to `application.html.erb` without having to worry about elements not being in the document by using the JQuery selector:

> Calling jQuery() (or $()) with an id selector as its argument will return a jQuery object containing a collection of either zero or one DOM element.
[-> Source](https://api.jquery.com/id-selector/)

It won't try to add an event listener to null (which is returned by .getElementById when the element is not found) and when .on() is called upon the object returned from $('#id') with zero DOM elements it tries to apply to the handler to each object in the collection, since there are no objects it "doesn't do" anything and continues on with the code, no error messages are shown in the console and we can include application.js directly :D

This was also useful to read:
http://api.jquery.com/on/
http://api.jquery.com/Types/#PlainObject
https://stackoverflow.com/questions/5724751/clarification-of-jquery-wrapper-object

fchavat replied:
That's much better! And jQuery will be useful for other things too! 💪

fchavat reviewed (commented):
Left some comments

ampeigonet committed:
Rewrite timer logic adressing comments

Fix clearinterval using an id + change several variables and hardcoded values to make code more reusable

ampeigonet committed:
Merge branch 'Uninstall-coffeescript-gem' into Start-timer-for-posts

retros => retro, minor refactors in css files

fchavat reviewed (approved):
LGTM

ampeigonet committed:
Merge branch 'master' into Start-timer-for-posts
